File: leetcode/0092_reverse_between.cpp
#include "./head.hpp"


// Definition for singly-linked list.
struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    ListNode* reverseBetween(ListNode* head, int left, int right) {
      ListNode* q1 = nullptr;
      int i = 1;
      while (i < left) {
        if (q1) {
          // not null
          q1 = q1->next;
        } else {
          // null
          q1 = head;
        }
        i++;
      }

      ListNode* p1 = nullptr;

      if (q1) {
        p1 = q1->next;
      } else {
        // null
        p1 = head;
      }

      ListNode* p2 = p1->next;
      while (i < right) {
        ListNode* t = p2->next;
        p2->next = p1;
        p1 = p2;
        p2 = t;
        i++;
      }

      if (q1) {
        ListNode* t = q1->next;
        q1->next = p1;
        t->next = p2;
        return head;
      } else {
        // null
        head->next = p2;
        return p1;
      }
    }
};

int main() {
  Solution s;
  ListNode* head = new ListNode(1, new ListNode(2, new ListNode(3, new ListNode(4, new ListNode(5)))));
  // ListNode* head = new ListNode(5);
  ListNode* p = s.reverseBetween(head, 5, 5);
  while (p) {
    std::cout << p->val << std::endl;
    p = p->next;
  }
  return 0;
}

File: leetcode/0092_reverse_between.rs
#[allow(dead_code)]
fn print_type_of<T>(_: &T) {
    println!("{}", std::any::type_name::<T>())
}

// Definition for singly-linked list.
#[derive(PartialEq, Eq, Clone, Debug)]
pub struct ListNode {
    pub val: i32,
    pub next: Option<Box<ListNode>>,
}

impl ListNode {
    #[inline]
    fn new(val: i32) -> Self {
        ListNode { next: None, val }
    }
}

struct Solution;

impl Solution {
    #[allow(unused_variables)]
    pub fn reverse_between(
        head: Option<Box<ListNode>>,
        left: i32,
        right: i32,
    ) -> Option<Box<ListNode>> {
        let mut head = head;
        let mut q1: &mut Option<Box<ListNode>> = &mut None;
        let mut i = 1i32;
        while i < left {
            match q1 {
                Some(qq1) => {
                    q1 = &mut qq1.next;
                }
                None => {
                    q1 = &mut head;
                }
            }
            i += 1;
        }
        let mut p1 = &mut None;
        if *q1 == None {
            p1 = &mut head;
        } else {
            p1 = &mut (*q1).as_mut().unwrap().next;
        }
        let mut p2 = &mut (*p1).as_mut().unwrap().next;
        while i < right {
            let t = &mut (*p2).as_mut().unwrap().next;
            (*p2).as_mut().unwrap().next = (*p1).clone();
            i += 1;
        }

        None
    }
}

fn main() {
    let mut head = Some(Box::new(ListNode::new(1)));
    head.as_mut().unwrap().next = Some(Box::new(ListNode::new(2)));
    head.as_mut().unwrap().next.as_mut().unwrap().next = Some(Box::new(ListNode::new(3)));
    head.as_mut()
        .unwrap()
        .next
        .as_mut()
        .unwrap()
        .next
        .as_mut()
        .unwrap()
        .next = Some(Box::new(ListNode::new(4)));
    println!("{:?}", Solution::reverse_between(head, 2, 3));
}
